tests/basic.rs

```diff
@@ -157,33 +167,63 @@ async fn test_stream_cancel_join() {
 
 #[tokio::test]
 async fn single_actor_on_address_with_stop_self_returns_disconnected_on_stop() {
-    let address = ActorReturningStopSelf.create(None).spawn_global();
+    let (address, ctx) = Context::new(None);
+    tokio::spawn(ctx.run(ActorStopSelf));
+    address.send(Stop).await.unwrap();
+    assert!(address
+        .join()
+        .timeout(Duration::from_secs(2))
+        .await
+        .is_some());
+}
 
+#[tokio::test]
+async fn two_actors_on_address_with_stop_self() {
+    let (address, mut ctx) = Context::new(None);
+    tokio::spawn(ctx.attach(ActorStopSelf));
+    tokio::spawn(ctx.run(ActorStopSelf));
     address.send(Stop).await.unwrap();
-    smol::Timer::after(Duration::from_secs(1)).await;
+    tokio::time::sleep(Duration::from_secs(1)).await;
 
+    assert!(address.is_connected());
+    address.send(Stop).await.unwrap();
+
+    tokio::time::sleep(Duration::from_secs(1)).await;
     assert!(!address.is_connected());
 }
 
-struct ActorReturningStopSelf;
+#[tokio::test]
+async fn two_actors_on_address_with_stop_self_context_alive() {
+    let (address, mut ctx) = Context::new(None);
+    tokio::spawn(ctx.attach(ActorStopSelf));
+    tokio::spawn(ctx.attach(ActorStopSelf)); // Context not dropped here
+    address.send(Stop).await.unwrap();
+    tokio::time::sleep(Duration::from_secs(1)).await;
+
+    assert!(address.is_connected());
+    address.send(Stop).await.unwrap();
+
+    tokio::time::sleep(Duration::from_secs(1)).await;
+    assert!(address.is_connected());
+}
```

Comment:
Is this really a good behaviour? I'd expect `address.is_connected()` to be equivalent to "can I expect that a message that I am about to send will be handled".

Reply:
Essentially, `is_connected` is equivalent to "can I expect that a message I am about to send **may** be handled". Since a free context still lives, this means that at a later time, another actor might be spawned onto that address (using `Context::run` or `Context::attach` and then handle the message.

Strictly speaking, **will** be handled was never a guarantee at all - the actor could stop itself with `Context::stop` before getting to a given message, even though when it was sent, `is_connected` returns true. 

> I'd expect address.is_connected() to be equivalent to "can I expect that a message that I am about to send will be handled".

I believe that this behaviour is currently used by [xtras](https://github.com/itchysats/itchysats/blob/master/xtras/src/supervisor.rs#L17) to implement a supervisor, though I haven't looked into it in great depth yet. The impression I got, though, was that if the address closes as soon as the last *actor* (as opposed to the last *context*) stops, then you can't restart an actor if it crashes using the supervisor pattern as currently implemented. Maybe there is another way to implement the supervisor pattern, though?

Reply:
> > I'd expect address.is_connected() to be equivalent to "can I expect that a message that I am about to send will be handled".
> 
> I believe that this behaviour is currently used by [xtras](https://github.com/itchysats/itchysats/blob/master/xtras/src/supervisor.rs?rgh-link-date=2022-06-14T20%3A59%3A25Z#L17) to implement a supervisor, though I haven't looked into it in great depth yet. The impression I got, though, was that if the address closes as soon as the last _actor_ (as opposed to the last _context_) stops, then you can't restart an actor if it crashes using the supervisor pattern as currently implemented. Maybe there is another way to implement the supervisor pattern, though?

Yep that is how I implemented it then but the implementation also guarantees that there won't ever be more than one actor running on this context.

> Essentially, `is_connected` is equivalent to "can I expect that a message I am about to send **may** be handled". Since a free context still lives, this means that at a later time, another actor might be spawned onto that address (using `Context::run` or `Context::attach` and then handle the message.

This is a bit of a surprising behaviour. I would expect `send` to return a `Disconnected` if no actor is alive and not have the message queued until a new actor is spawned, just because the `Context` is still there? Also, is this even true if a combination of `attach` and `run` is used? I guess as soon as you use `run` once, the context will be gone with the last actor stopping.

Reply:
> Yep that is how I implemented it then but the implementation also guarantees that there won't ever be more than one actor running on this context.

That is true. On the other hand, it's probably a bit counterintuitive to make it shut down the address if 2/2 actors on it stop, versus not shutting down the address if 1/1 actors on it stop.

> I would expect send to return a Disconnected if no actor is alive and not have the message queued until a new actor is spawned.

I am mostly worried that this would send the wrong signal to users that "this address is dead, we can throw it away", which would not be the case.

> just because the Context is still there? 

Most of the time, when an actor stops, the only contexts are running. I think that the only exception to this would be in the case of the supervisor pattern or another time when you specifically relying on this behaviour. Therefore, it probably isn't so surprising from either the supervisor or the sender side - from the supervisor's perspective, the implementation becomes impossible if this behaviour is removed. From the sender side, since it may be restarted to handle a message later, `Disconnected` early gives the wrong impression.

> I guess as soon as you use `run` once, the context will be gone with the last actor stopping.

That is true. Technically, though, you could `attach`, say, 2 actors, and when they stop `run`, consuming the context. I'm not sure why anyone would ever do this, though.

Reply:
So, do you think we can merge it? If this behaviour really shouldn't exist at all, we can discuss this further in another PR removing it, I think, since the only change here is that now `stop_all` does the same thing as `stop_self` run on all actors. Fixing this inconsistency I personally think is a bugfix, as it was counterintuitive and undocumented.

Reply:
Yes. I think the behaviour also makes sense with the current capabilities exposed by `xtra`, i.e. the `attach` method.

Slight OT: I am wondering if there is a better way to implement multiple actors behind one address. Perhaps once we get the inbox reworkings in, creating multiple event loops listening on the same address can be implemented on top and thus with different semantics!

Reply:
> Perhaps once we get the inbox reworkings in, creating multiple event loops listening on the same address can be implemented on top and thus with different semantics!

Once it's done, I'd be interested in seeing how this could be achieved! My first thought is that it might depend rather heavily on the inbox channel semantics, though